src/core/label_provider.h

```diff
@@ -38,17 +38,28 @@ class LabelProvider {
  public:
   LabelProvider() = default;
 
+  using Labels = std::vector<std::string>;
+
   // Return the label for a given 'index' of a 'name'. Return empty
   // string if no label is available.
   const std::string& GetLabel(const std::string& name, size_t index) const;
 
+  // Return the labels for a given 'name'. Return nullptr
```

Comment:
LabelProvider is only the labels, so it doesn't make sense to just shared_ptr the labels part. Change backend.h to use a shared_ptr for LabelProvider there and just use shared_ptr<LabelProvider> everywhere to share it. I don't think LabelProvider needs any changes (except the header comments need updating... I actually have a pending change for that).